torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -0,0 +1,259 @@
+    } else {
+      printf("\nno reduction\n");
+      // Merge all dimensions because we're only supporting pointwise
+      while (out->nDims() > 1)
+        out->merge(0, 1);
+    }
+    printf("\ndone output one\n");
+  }
+
+  printf("\nstep 1\n");
+
+  if (has_reduction) {
+    // Run through outputs, grab all inputs of outputs
+    // squeeze with computeAt to set overall structure.
+    for (auto output : fusion->outputs()) {
+      if (output->getValType() != ValType::TensorView)
+        continue;
+      TensorView* out_tv = static_cast<TensorView*>(output);
+
+      // fcd_reduction could be queried later via
+      // cuda_kernel_->reduction_axes_, which would ensure we have proper
+      // launch configuratoin.
+      TensorView* intermediate;
+      if (fcd_reduction) {
+        out_tv->split(-1, kFcdReductionThreadX);
+        // necessary to avoid dynamic allocation on intermediates;
+        intermediate = out_tv->rFactor({-2});
+      } else {
+        // TODO: we don't need a full warp here, this should be determined by
+        //       element data type
+        out_tv->split(0, kNonFcdReductionThreadX);
+        out_tv->split(
+            -1, kNonFcdReductionThreadY); // necessary to avoid dynamic
+                                          // allocation on intermediates;
+        intermediate = out_tv->rFactor({-2});
+      }
+      for (Val* inp : fusion->inputsOf(output)) {
+        // scheduling of inputs shouldn't change with different fcd_reduction
+        if (inp->getValType().value() == ValType::TensorView) {
+          static_cast<TensorView*>(inp)->computeAt(intermediate, -1);
+        }
+      }
+      // scheduling of inputs shouldn't change with different fcd_reduction
+      intermediate->computeAt(out_tv, -2);
+      if (fcd_reduction) {
+        out_tv->axis(0)->parallelize(ParallelType::BIDx);
+      } else {
+        out_tv->axis(0)->parallelize(ParallelType::BIDx);
+        out_tv->axis(1)->parallelize(ParallelType::TIDx);
+      }
+    }
+    // Run through all values, unroll, and bind their axes
+    for (auto val : fusion->vals()) {
+      if (val->getValType().value() != ValType::TensorView)
+        continue;
+      TensorView* tv = static_cast<TensorView*>(val);
+      if (fcd_reduction) {
+        tv->axis(-1)->parallelize(ParallelType::TIDx);
+      } else {
+        tv->axis(-1)->parallelize(ParallelType::TIDy);
+      }
+    }
+
+    TensorView* out0 = fusion->output(0)->as<TensorView>();
+    int ndim = (int)out0->nDims();
+    Val* numel = new Int(1);
+    std::cout << "\ndebugging output\n" << out0 << std::endl;
+    std::cout << std::endl << fusion << std::endl;
+    for (int i = 0; i < ndim; i++) {
+      if (out0->axis(i)->isBlockDim()) {
+        numel = mul(numel, out0->axis(i)->rawExtent());
+      }
+    }
+    if (fcd_reduction) {
+      // assuming all output to be the same shape;
+      fusion->setLaunchConfig(LaunchConfigType::TIDx, new Int(kFcdReductionThreadX));
```

Comment:
if we always map integers why not have the launch config map to a Int::ScalarType instead of a Int* ?

Reply:
Are we suggesting to have the `ExpressionEvaluator::evaluate` wrapped inside this as well? We still need to pass a `Int*`/`Val*` through the API, as later `ExpressionEvaluator::evaluator(Val* val, const EvaluationContext* context)` takes argument `Val*`.

Reply:
Discussed offline. launch config should be inferred at run-time, so we need pointer to nodes.